File: data_structures/LinkedList/LinkedList.java
/*
    Linked List implementation, as of 5 April, 2018.

    Background: *some of this information is from the Linked List wikipedia page
        IMO one of the ancestral data structures that make up other data 
        structures. 

        Advantages:
        - Relocatable; can be moved about in memory at will, and it can also be
          quickly and directly serialized for storage on disk or transfer over
          a network.
        - Especially for a small list, array indexes can occupy significantly
          less space than a full pointer on many architectures.
        - Using an entry from a pre-allocated array is faster than using 
          dynamic memory allocation for each node, since dynamic memory 
          allocation typically requires a search for a free memory block of the
          desired size.

        Disadvantages:
        - Linked lists use more memory than arrays bc of storing pointers.
        - Must be searched in order; no indexing technically.
        - Nodes are stored incontiguously, greatly increasing time to access 
          elements
        - Reverse traversal is cumbersome if not doubly linked. If doubly
          linked, memory is consumed allocating space for a back pointer.

    Big O:
        Indexing: O(n)
        Insert/delete at beginning: O(1)
        Insert/delete at end: O(1) when last element known; O(n) when not known
        Insert/delete in middle: Search Time + O(1)
        Wasted Space (average): O(n)

    Uses:
        1. When you need O(1) insert/delete from list
        2. When you don't know how many items at runtime
        3. You don't need random access to any elements
        4. You want to be able to insert items in the middle of the list

    TODO:      
        how to reverse a ll? 
*/
class LinkedList  
{   
    Node head; // head node

    static class Node
    {
        int data;
        Node next;

        Node(int d) { data = d; next = null; }
    }

    public void push(int new_data)
    {
        Node new_node = new Node(new_data);
        new_node.next = head;
        head = new_node;
    }

    public void append(int new_data)
    {
        Node new_node = new Node(new_data);

        if(head == null)
        {
            head = new Node(new_data);
            return;
        }

        new_node.next = null;

        Node last = head;
        while(last.next != null)
        {
            last = last.next;
        }

        last.next = new_node;
    }

    public void printList()
    {
        Node n = head;
        while(n != null)
        {
            System.out.print(n.data + " ");
            n = n.next;
        }
        System.out.println();
    }

    public void reverseList()
    {
        // create new list, reversedList
        LinkedList reversedList = new LinkedList();

        // create tmp node that will contain the last node information
        Node last = head;
        Node tmp = null;

        // go to last node with node.next = null 
        while(last.next != null)
        {
            tmp = last;
            last = last.next;
        }

        System.out.println("last.data: " + last.data);
        System.out.println("last: " + last);

        System.out.println(tmp + "    " + last);
        if(tmp == last)
        {
            System.out.println("tmp == last; true");
        }
        else
        {
            System.out.println("false");
        }
        // save node pointer as next
        // save node pointer as next
    }

    public static void main(String[] args)
    {
        LinkedList l = new LinkedList();

        l.head = new Node(1);
        Node second = new Node(2);
        Node third = new Node(3);

        l.head.next = second;
        second.next = third;

        l.printList();

        l.reverseList();

        l.printList();
    }
}

File: data_structures/LinkedList/linkedList.cpp
#include <iostream>
using namespace std;

/* 8.1.18
To remember:
    - Arrays can be used to store linear data of similar types, but have limitations:
        1. Size of arrays is fixed; must know upper bound of elements. Also memory is pre-allocated.
        2. Inserting a new ele in array is expensive, need to create new array.
        3. Deletion is also expensive with arrays unless use special techniques.
    - Pros of linked list
        1. Dynamic size
        2. Ease of insertion/deletion
    - Cons of linked list
        1. Random access not allowed. Need to access elements sequentially.
        2. Extra memory space for a pointer is required with each element of list.
        3. Not cache friendly. Not contiguous memory locations.
*/
struct Node
{
    int data;
    struct Node *next;
};

// O(n) to print bc iterating through linked list and printing each
void printList(struct Node *head)
{
    while(head != NULL)
    {
        cout << head->data << endl;
        head = head->next;
    }
}

// O(1) to push because pushing to head
void pushAtHead(struct Node **head, int new_data)
{
    struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));

    new_node->data = new_data;
    new_node->next = *head;
    *head = new_node;
}

// O(1) to insert after node because node is given to function
void insertAfter(struct Node *prev, int new_data)
{
    if(prev == NULL)
    {
        cout << "Given previous node cannot be NULL" << endl;
        return;
    }

    struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));

    new_node->data = new_data;
    new_node->next = prev->next;
    prev->next = new_node;
}

// O(n) because iterate through list; could include tail node to reduce to O(1)
void pushAtTail(struct Node **head, int new_data)
{
    struct Node *new_node = (struct Node*)malloc(sizeof(struct Node));
    struct Node *last = *head;

    new_node->data = new_data;
    new_node->next = NULL;

    if(*head == NULL)
    {
        *head = new_node;
        return;
    }

    while(last->next != NULL)
    {
        last = last->next;
    }

    last->next = new_node;
}

int main()
{
    struct Node* head   = NULL;
    struct Node* second = NULL;
    struct Node* third  = NULL;

    head = (struct Node*)malloc(sizeof(struct Node));
    second  = (struct Node*)malloc(sizeof(struct Node));
    third = (struct Node*)malloc(sizeof(struct Node));

    head->data = 1;
    head->next = second;

    second->data = 2;
    second->next = third;

    third->data = 3;
    third->next = NULL;

    printList(head);

    pushAtHead(&head, 0);
    printList(head);

    pushAtTail(&head, 10);
    printList(head);
    pushAtTail(&head, 10);
}
